066-OpenAIFundamentals/Student/Resources/Notebooks/CH-03-B-Chunking.ipynb

```diff
@@ -0,0 +1,402 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "! pip install --upgrade click\n",
+    "! python -m spacy download en_core_web_sm"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import openai\n",
+    "import PyPDF3\n",
+    "import os\n",
+    "import json\n",
+    "import tiktoken\n",
+    "import spacy\n",
+    "from openai.error import InvalidRequestError\n",
+    "\n",
+    "from dotenv import load_dotenv, find_dotenv\n",
+    "load_dotenv(find_dotenv())\n",
+    "\n",
+    "from spacy.lang.en import English \n",
+    "nlp = spacy.load(\"en_core_web_sm\")\n",
+    "\n",
+    "import langchain\n",
+    "from langchain.text_splitter import RecursiveCharacterTextSplitter"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Set up your environment to access your Azure OpenAI keys. Refer to your Azure OpenAI resource in the Azure Portal to retrieve information regarding your Azure OpenAI endpoint and keys. \n",
+    "\n",
+    "For security purposes, store your sensitive information in an .env file."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Load your OpenAI credentials\n",
+    "API_KEY = os.getenv(\"OPENAI_API_KEY\")\n",
+    "assert API_KEY, \"ERROR: Azure OpenAI Key is missing\"\n",
+    "openai.api_key = API_KEY\n",
+    "\n",
+    "RESOURCE_ENDPOINT = os.getenv(\"OPENAI_API_BASE\",\"\").strip()\n",
+    "assert RESOURCE_ENDPOINT, \"ERROR: Azure OpenAI Endpoint is missing\"\n",
+    "assert \"openai.azure.com\" in RESOURCE_ENDPOINT.lower(), \"ERROR: Azure OpenAI Endpoint should be in the form: \\n\\n\\t<your unique endpoint identifier>.openai.azure.com\"\n",
+    "openai.api_base = RESOURCE_ENDPOINT\n",
+    "\n",
+    "openai.api_type = os.getenv(\"OPENAI_API_TYPE\")\n",
+    "openai.api_version = os.getenv(\"OPENAI_API_VERSION\")\n",
+    "model=os.getenv(\"CHAT_MODEL_NAME\")\n"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## 3. Counting Tokens\n",
+    "\n",
+    "Tiktoken uses a technique called BPE, or byte pair encoding to convert the given text into tokens. There are different encodings available to help process the words. In this notebook, we will use the cl100k_base.\n",
+    "\n",
+    "#### Student Challenge #1: \n",
+    "\n",
+    "Count the number of tokens in the final answer we received in CH-03-A-Grounding by completing the function, count_tokens, below. "
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "\n",
+    "def count_tokens(string: str, encoding_name: str) -> int:\n",
+    "    # Enter code here\n",
+    "    return "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### Student Challenge #2:\n",
```

Comment:
Note: this comment applies generally to this file. Within this challenge, there are multiple challenges which is a bit confusing. Perhaps the challenges within this file should be named "tasks" instead.